File: sdk/services/sandbox/default-policy.ts
/**
 * Default Sandbox Policy
 *
 * Provides safe default values for all sandbox sub-policies.
 * Architecture reference: docs/infra/sandbox/architecture.md
 */

import type {
  SandboxPolicy,
  ShellPolicy,
  PythonPolicy,
  JavaScriptPolicy,
  LuaPolicy,
  FilesystemPolicy,
  ProcessPolicy,
  NetworkPolicy,
  ResourcePolicy,
} from "@wf-agent/types";

export const DEFAULT_SHELL_POLICY: ShellPolicy = {
  allowedCommands: [],
  deniedCommands: [
    "sudo",
    "su",
    "chroot",
    "mount",
    "umount",
    "dd",
    "mkfs",
    "mkfs.ext4",
    "mkfs.xfs",
    "reboot",
    "shutdown",
    "poweroff",
    "halt",
    "passwd",
    "useradd",
    "userdel",
    "usermod",
    "groupadd",
    "groupdel",
    "chown",
    "chmod",
    "insmod",
    "rmmod",
    "modprobe",
  ],
  dangerousPatterns: [
    "rm\\s+(-rf|--recursive)\\s+\\/\\s*",
    ":\\(\\)\\s*\\{.*:\\(\\)\\s*\\}.*\\}",
    "\\|\\s*(bash|sh|zsh)\\s",
    "curl.*\\|.*(bash|sh)",
    "wget.*\\|.*(bash|sh)",
    "LD_PRELOAD=",
    "PYTHONPATH=.*:",
    "\\/dev\\/(null|zero|random|urandom)",
    "fork\\s*\\(\\)\\s*\\{",
    "exec\\s+\\/s?bin\\/(init|systemd)",
  ],
  allowPipe: true,
  allowRedirect: true,
};

export const DEFAULT_PYTHON_POLICY: PythonPolicy = {
  allowedModules: [],
  deniedModules: [
    "os",
    "subprocess",
    "shutil",
    "ctypes",
    "socket",
    "pty",
    "signal",
    "multiprocessing",
    "distutils",
    "sysconfig",
  ],
  allowSubprocess: false,
  restrictBuiltinOpen: true,
  allowDynamicEval: false,
};

export const DEFAULT_JS_POLICY: JavaScriptPolicy = {
  allowedModules: [],
  deniedModules: [
    "child_process",
    "cluster",
    "worker_threads",
    "v8",
    "vm",
    "inspector",
    "module",
    "process",
  ],
  allowChildProcess: false,
  allowFSWrite: false,
  allowDynamicEval: false,
};

export const DEFAULT_LUA_POLICY: LuaPolicy = {
  allowedModules: [],
  deniedModules: ["os", "io", "package", "debug", "ffi", "socket", "lfs", "luaposix"],
  allowOsExecute: false,
  restrictIoOpen: true,
  allowDynamicLoad: false,
};

export const DEFAULT_FILESYSTEM_POLICY: FilesystemPolicy = {
  allowedReadPaths: [],
  allowedWritePaths: [],
  allowedRemovePaths: [],
  allowedExecutePaths: [],
  copyOnWrite: false,
  maxFileSize: 10 * 1024 * 1024,
};

export const DEFAULT_PROCESS_POLICY: ProcessPolicy = {
  allowedChildProcesses: [],
  deniedChildProcesses: [],
  maxChildProcesses: 0,
  allowFork: false,
  allowExec: false,
};

export const DEFAULT_NETWORK_POLICY: NetworkPolicy = {
  access: "none",
  allowDns: false,
};

export const DEFAULT_RESOURCE_POLICY: ResourcePolicy = {
  timeoutLimit: 30000,
};

export const DEFAULT_SANDBOX_POLICY: SandboxPolicy = {
  mode: "strict",
  shell: DEFAULT_SHELL_POLICY,
  python: DEFAULT_PYTHON_POLICY,
  javascript: DEFAULT_JS_POLICY,
  lua: DEFAULT_LUA_POLICY,
  filesystem: DEFAULT_FILESYSTEM_POLICY,
  process: DEFAULT_PROCESS_POLICY,
  network: DEFAULT_NETWORK_POLICY,
  resource: DEFAULT_RESOURCE_POLICY,
};

/**
 * Shell Policy Presets for different security requirements.
 *
 * SAFE:     Strict mode — only whitelist commands allowed, no pipe/redirect
 * BALANCED: Moderate mode — deny high-risk commands, allow pipe, deny redirect
 * PERMISSIVE: Loose mode — only deny most dangerous commands, allow pipe and redirect
 */
export const SHELL_POLICY_PRESETS = {
  SAFE: {
    allowedCommands: [
      "git",
      "npm",
      "pnpm",
      "node",
      "ls",
      "cat",
      "echo",
      "cd",
      "pwd",
      "mkdir",
      "cp",
      "mv",
      "rm",
    ],
    deniedCommands: DEFAULT_SHELL_POLICY.deniedCommands,
    dangerousPatterns: DEFAULT_SHELL_POLICY.dangerousPatterns,
    allowPipe: false,
    allowRedirect: false,
  },
  BALANCED: {
    allowedCommands: [],
    deniedCommands: DEFAULT_SHELL_POLICY.deniedCommands,
    dangerousPatterns: DEFAULT_SHELL_POLICY.dangerousPatterns,
    allowPipe: true,
    allowRedirect: false,
  },
  PERMISSIVE: {
    allowedCommands: [],
    deniedCommands: ["sudo", "su", "chroot", "dd", "mkfs", "reboot", "shutdown"],
    dangerousPatterns: ["rm\\s+(-rf|--recursive)\\s+\\/\\s*"],
    allowPipe: true,
    allowRedirect: true,
  },
} as const satisfies Record<string, ShellPolicy>;

File: crates/wf-sandbox/src/default_policy.rs
use std::sync::OnceLock;

use wf_types::script::sandbox::{
    FilesystemPolicy, JavaScriptPolicy, LuaPolicy, NetworkAccessType, NetworkPolicy, ProcessPolicy,
    PythonPolicy, ResourcePolicy, SandboxMode, SandboxPolicy, ShellPolicy,
};

fn build_default_sandbox_policy() -> SandboxPolicy {
    SandboxPolicy {
        mode: Some(SandboxMode::Strict),
        shell: Some(ShellPolicy {
            allowed_commands: None,
            denied_commands: Some(vec![
                "sudo".to_string(),
                "su".to_string(),
                "chroot".to_string(),
            ]),
            dangerous_patterns: Some(vec![
                "rm\\s+(-rf|--recursive)".to_string(),
                ":\\(\\)\\s*\\{.*:\\(\\)\\s*\\}.*\\}".to_string(),
            ]),
            allow_pipe: Some(true),
            allow_redirect: Some(true),
        }),
        python: Some(PythonPolicy {
            allowed_modules: Some(vec![]),
            denied_modules: Some(vec!["os".to_string(), "subprocess".to_string()]),
            allow_subprocess: Some(false),
            restrict_builtin_open: Some(true),
            allow_dynamic_eval: Some(false),
        }),
        javascript: Some(JavaScriptPolicy {
            allowed_modules: Some(vec![]),
            denied_modules: Some(vec!["child_process".to_string(), "fs".to_string()]),
            allow_child_process: Some(false),
            allow_fs_write: Some(false),
            allow_dynamic_eval: Some(false),
        }),
        lua: Some(LuaPolicy {
            allowed_modules: Some(vec![]),
            denied_modules: Some(vec![
                "os".to_string(),
                "io".to_string(),
                "package".to_string(),
                "debug".to_string(),
                "ffi".to_string(),
            ]),
            allow_os_execute: Some(false),
            restrict_io_open: Some(true),
            allow_dynamic_load: Some(false),
        }),
        filesystem: Some(FilesystemPolicy {
            allowed_read_paths: Some(vec![]),
            allowed_write_paths: Some(vec![]),
            allowed_remove_paths: Some(vec![]),
            allowed_execute_paths: Some(vec![]),
            copy_on_write: Some(true),
            max_file_size: Some(10 * 1024 * 1024),
        }),
        process: Some(ProcessPolicy {
            allowed_child_processes: Some(vec![]),
            denied_child_processes: Some(vec![]),
            max_child_processes: Some(10),
            // Exec/fork are permitted by default so the default chain
            // (static-analyzer + os-hook) stays usable for external commands;
            // the analysis gate enforces command-level policy and the seccomp
            // layer denies them when a user explicitly sets these to false.
            allow_fork: Some(true),
            allow_exec: Some(true),
        }),
        network: Some(NetworkPolicy {
            access_type: Some(NetworkAccessType::None),
            allowed_domains: None,
            allowed_ports: None,
            allow_dns: Some(false),
        }),
        resource: Some(ResourcePolicy {
            cpu_limit_ms: None,
            memory_limit_mb: Some(512),
            disk_limit_mb: Some(1024),
            timeout_limit_ms: Some(30000),
        }),
    }
}

pub static DEFAULT_SANDBOX_POLICY: OnceLock<SandboxPolicy> = OnceLock::new();

pub fn default_sandbox_policy() -> &'static SandboxPolicy {
    DEFAULT_SANDBOX_POLICY.get_or_init(build_default_sandbox_policy)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_policy_is_strict() {
        let policy = default_sandbox_policy();
        assert_eq!(policy.mode, Some(SandboxMode::Strict));
    }

    #[test]
    fn test_default_policy_permits_exec_and_fork() {
        let policy = default_sandbox_policy();
        let process = policy.process.as_ref().unwrap();
        assert_eq!(process.allow_exec, Some(true));
        assert_eq!(process.allow_fork, Some(true));
    }

    #[test]
    fn test_default_policy_denies_sudo() {
        let policy = default_sandbox_policy();
        let shell = policy.shell.as_ref().unwrap();
        assert!(shell
            .denied_commands
            .as_ref()
            .unwrap()
            .contains(&"sudo".to_string()));
    }

    #[test]
    fn test_default_policy_lua_denies_os() {
        let policy = default_sandbox_policy();
        let lua = policy.lua.as_ref().unwrap();
        assert!(lua
            .denied_modules
            .as_ref()
            .unwrap()
            .contains(&"os".to_string()));
    }
}
